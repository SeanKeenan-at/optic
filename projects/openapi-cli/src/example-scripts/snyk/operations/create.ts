import { commonHeaders, commonResponses, refs } from '../common';
import { OpenAPIV3 } from 'openapi-types';
import {
  buildCreateRequestSchema,
  buildItemResponseSchema,
  ensureRelationSchema,
} from '../schemas';

export function addCreateOperation(
  spec: OpenAPIV3.Document,
  collectionPath: string,
  resourceName: string,
  titleResourceName: string
): void {
  if (!spec.paths) spec.paths = {};
  if (!spec.paths[collectionPath]) spec.paths[collectionPath] = {};
  if (!spec.components) spec.components = {};
  if (!spec.components.schemas) spec.components.schemas = {};
  spec.paths[collectionPath]!.post = buildCreateOperation(
    resourceName,
    titleResourceName
  );
  const attributes =
    spec.components?.schemas?.[`${titleResourceName}Attributes`];
  if (!attributes)
    throw new Error(`Could not find ${titleResourceName}Attributes schema`);
  spec.components.schemas[`${titleResourceName}CreateAttributes`] = attributes;
  ensureRelationSchema(spec, titleResourceName);
}

function buildCreateOperation(
  resourceName: string,
  titleResourceName: string
): OpenAPIV3.OperationObject {
  const itemResponseSchema = buildItemResponseSchema(
    resourceName,
    titleResourceName
  );
  const createRequestSchema = buildCreateRequestSchema(titleResourceName);
  return {
    summary: `Create a new ${resourceName}`,
    description: `Create a new ${resourceName}`,
    operationId: `create${titleResourceName}`,
    tags: [titleResourceName],
    parameters: [refs.parameters.version],
    requestBody: {
      content: {
        'application/json': {
          schema: createRequestSchema,
        },
      },
    },
    responses: {
      '201': {
        description: `Created ${resourceName} successfully`,
        headers: commonHeaders,
        content: {
          'application/vnd.api+json': {
            schema: itemResponseSchema,
          },
        },
      },
      ...commonResponses,
    },
  };
}
