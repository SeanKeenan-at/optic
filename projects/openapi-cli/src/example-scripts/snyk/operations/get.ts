import { ensureIdParameter } from '../parameters';
import { buildItemResponseSchema, ensureRelationSchema } from '../schemas';
import { commonHeaders, commonResponses, refs } from '../common';
import { OpenAPIV3 } from 'openapi-types';

export function addGetOperation(
  spec: OpenAPIV3.Document,
  itemPath: string,
  resourceName: string,
  titleResourceName: string
): void {
  if (!spec.paths) spec.paths = {};
  if (!spec.paths[itemPath]) spec.paths[itemPath] = {};
  spec.paths[itemPath]!.get = buildGetOperation(
    resourceName,
    titleResourceName
  );
  ensureIdParameter(spec, resourceName, titleResourceName);
  ensureRelationSchema(spec, titleResourceName);
}

function buildGetOperation(
  resourceName: string,
  titleResourceName: string
): OpenAPIV3.OperationObject {
  const itemResponseSchema = buildItemResponseSchema(
    resourceName,
    titleResourceName
  );
  return {
    summary: `Get instance of ${resourceName}`,
    description: `Get instance of ${resourceName}`,
    operationId: `get${titleResourceName}`,
    tags: [titleResourceName],
    parameters: [
      refs.parameters.version,
      { $ref: `#/components/parameters/${titleResourceName}Id` },
    ],
    responses: {
      '200': {
        description: `Returns an instance of ${resourceName}`,
        headers: commonHeaders,
        content: {
          'application/vnd.api+json': {
            schema: itemResponseSchema,
          },
        },
      },
      ...commonResponses,
    },
  };
}
