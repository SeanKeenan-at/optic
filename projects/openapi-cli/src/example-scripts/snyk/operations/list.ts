import { OpenAPIV3 } from 'openapi-types';
import {
  buildCollectionResponseSchema,
  ensureRelationSchema,
} from '../schemas';
import {
  commonHeaders,
  commonResponses,
  paginationParameters,
  refs,
} from '../common';

export function addListOperation(
  spec: OpenAPIV3.Document,
  collectionPath: string,
  resourceName: string,
  titleResourceName: string
): void {
  if (!spec.paths) spec.paths = {};
  if (!spec.paths[collectionPath]) spec.paths[collectionPath] = {};
  spec.paths[collectionPath]!.get = buildListOperation(
    resourceName,
    titleResourceName
  );
  ensureRelationSchema(spec, titleResourceName);
}

function buildListOperation(
  resourceName: string,
  titleResourceName: string
): OpenAPIV3.OperationObject {
  const collectionResponseSchema = buildCollectionResponseSchema(
    resourceName,
    titleResourceName
  );
  return {
    summary: `List instances of ${resourceName}`,
    description: `List instances of ${resourceName}`,
    operationId: `list${titleResourceName}`,
    tags: [titleResourceName],
    parameters: [refs.parameters.version, ...paginationParameters],
    responses: {
      '200': {
        description: `Returns a list of ${resourceName} instances`,
        headers: commonHeaders,
        content: {
          'application/vnd.api+json': {
            schema: collectionResponseSchema,
          },
        },
      },
      ...commonResponses,
    },
  };
}
