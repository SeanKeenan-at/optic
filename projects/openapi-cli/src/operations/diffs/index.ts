import { Operation } from '..';
import { CapturedInteraction } from '../../captures';
import { OperationDiffResult } from './result';
import { OperationDiffTraverser } from './traverser';

export type { OperationDiffResult };

export function* diffInteractionByOperation(
  interaction: CapturedInteraction,
  operation: Operation
): IterableIterator<OperationDiffResult> {
  const traverser = new OperationDiffTraverser();
  traverser.traverse(interaction, operation);
  yield* traverser.results();
}
