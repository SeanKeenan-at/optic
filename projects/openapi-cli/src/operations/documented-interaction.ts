import { Operation } from '.'

export interface DocumentedInteraction {
  interaction: CapturedInteraction;
  operation: Operation;
  specJsonPath: string;
}

export class Documented