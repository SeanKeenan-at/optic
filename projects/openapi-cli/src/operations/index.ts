import { CapturedInteraction } from '../captures';
import { OpenAPIV3 } from '../specs';
import { OperationLocation } from '@useoptic/openapi-utilities';
import { SchemaObject } from '../shapes';

export type Operation = OpenAPIV3.OperationObject & {
  path: string;
  method: OpenAPIV3.HttpMethods;
};

export interface DocumentedInteraction {
  interaction: CapturedInteraction;
  operation: Operation;
  specJsonPath: string;
}
