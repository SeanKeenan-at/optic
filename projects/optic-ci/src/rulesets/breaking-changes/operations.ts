import { ApiChangeDsl, OpenAPIV3 } from '@useoptic/api-checks';
import { expect } from 'chai';

export const rules = {
  removingOperationId: ({ operations }: ApiChangeDsl) => {
    operations.changed.must(
      'have consistent operation IDs',
      (current, next) => {
        expect(current.operationId).to.equal(next.operationId);
      }
    );
  },
  preventRemovingOperation: ({ operations }: ApiChangeDsl) => {
    operations.removed.must('not be allowed', (operation, context) => {
      expect.fail('expected operation to be present');
    });
  },
  preventAddingRequiredQueryParameters: ({ request }: ApiChangeDsl) => {
    request.queryParameter.added.must('not be required', (queryParameter) => {
      expect(queryParameter.required).to.not.be.true;
    });
  },
  preventChangingOptionalToRequiredQueryParameters: ({
    request,
  }: ApiChangeDsl) => {
    request.queryParameter.changed.must(
      'not be optional then required',
      (queryParameterBefore, queryParameterAfter) => {
        if (!queryParameterBefore.required) {
          expect(queryParameterAfter.required).to.not.be.true;
        }
      }
    );
  },
  preventRemovingStatusCodes: ({ responses }: ApiChangeDsl) => {
    responses.removed.must('not be removed', (response) => {
      expect(false, `expected ${response.statusCode} to be present`).to.be.true;
    });
  },
  preventChangingParameterDefaultValue: ({ request }: ApiChangeDsl) => {
    request.queryParameter.changed.must(
      'not change the default value',
      (parameterBefore, parameterAfter) => {
        let beforeSchema = (parameterBefore.schema ||
          {}) as OpenAPIV3.SchemaObject;
        let afterSchema = (parameterAfter.schema ||
          {}) as OpenAPIV3.SchemaObject;
        expect(beforeSchema.default).to.equal(afterSchema.default);
      }
    );
  },
};
