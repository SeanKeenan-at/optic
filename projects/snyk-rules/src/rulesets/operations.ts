import { SnykApiCheckDsl } from "../dsl";
const chai = require("chai");
const { expect } = chai;
// @ts-ignore
chai.use(function (_chai, _) {
  // @ts-ignore
  _chai.Assertion.addMethod("withMessage", function (msg) {
    // @ts-ignore
    _.flag(this, "message", msg);
  });
});

export const rules = {
  operationId: ({ operations }: SnykApiCheckDsl) => {
    operations.requirement.must(
      "have the correct operationId format",
      (operation) => {
        expect(operation.operationId).to.be.ok;
        const camelCaseRe = /^(get|create|list|update|delete)([A-Z][a-z]+)+$/g;
        expect(
          camelCaseRe.test(operation.operationId || ""),
          `operationId "${
            operation.operationId || ""
          }" must be formatted [get|create|list|update|delete]camelCaseResource`
        ).to.be.ok;
      }
    );
  },
  tags: ({ operations }: SnykApiCheckDsl) => {
    operations.requirement.must("have tags", (operation) => {
      expect(operation.tags).withMessage(
        "tags [] must be set on each operation"
      ).to.exist;
      expect(operation.tags)
        .withMessage("tags [] must have at least one tag")
        .to.have.lengthOf.above(0);
    });
  },
  summary: ({ operations }: SnykApiCheckDsl) => {
    operations.requirement.must("have a summary", (operation) => {
      expect(operation.summary).to.exist;
    });
  },
};
