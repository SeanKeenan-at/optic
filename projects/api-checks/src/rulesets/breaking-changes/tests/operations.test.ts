import { rules } from '../operations';

import { createTestFixture } from './fixtures';
const { compare } = createTestFixture();

describe('operationId', () => {
  const baseForOperationIdTests = {
    openapi: '3.0.1',
    paths: {
      '/example': {
        get: {
          responses: {},
        },
      },
    },
    info: { version: '0.0.0', title: 'Empty' },
  };

  it('fails if removed', async () => {
    const baseCopy = JSON.parse(JSON.stringify(baseForOperationIdTests));
    baseCopy.paths['/example'].get.operationId = 'example';
    const result = await compare(baseCopy)
      .to((spec) => {
        delete spec.paths!['/example']!.get!.operationId;
        return spec;
      })
      .withRule(rules.removingOperationId, {});

    expect(result.results[0].passed).toBeFalsy();
    expect(result).toMatchSnapshot();
  });

  it('fails if changed', async () => {
    // todo: fix copy/paste
    const baseCopy = JSON.parse(JSON.stringify(baseForOperationIdTests));
    baseCopy.paths['/example'].get.operationId = 'example';
    const result = await compare(baseCopy)
      .to((spec) => {
        spec.paths!['/example']!.get!.operationId = 'example2';
        return spec;
      })
      .withRule(rules.removingOperationId, {});

    expect(result.results[0].passed).toBeFalsy();
    expect(result).toMatchSnapshot();
  });
});

const baseForOperationMetadataTests = {
  openapi: '3.0.1',
  paths: {
    '/example': {
      get: {
        tags: ['Example'],
        operationId: 'getExample',
        summary: 'Retrieve example',
        responses: {},
      },
    },
  },
  info: { version: '0.0.0', title: 'Empty' },
};

describe('operation parameters', () => {
  describe('status codes', () => {
    it('fails when a status codes is removed', async () => {
      const base = JSON.parse(JSON.stringify(baseForOperationMetadataTests));
      base.paths['/example'].get.responses = {
        '200': {
          description: 'Example response',
        },
      };
      const result = await compare(base)
        .to((spec) => {
          delete spec.paths!['/example']!.get!.responses!['200'];
          return spec;
        })
        .withRule(rules.preventRemovingStatusCodes, {});

      expect(result.results[0].passed).toBeFalsy();
      expect(result).toMatchSnapshot();
    });
  });
});
