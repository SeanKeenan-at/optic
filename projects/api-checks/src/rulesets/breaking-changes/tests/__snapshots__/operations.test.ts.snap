// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`operation parameters status codes fails when a status codes is removed 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "operationId": "getExample",
          "responses": Object {
            "200": Object {
              "description": "Example response",
            },
          },
          "summary": "Retrieve example",
          "tags": Array [
            "Example",
          ],
        },
      },
    },
  },
  "changelog": Array [
    Object {
      "changeType": "removed",
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "statusCode": "200",
          },
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
          "responses",
          "200",
        ],
        "jsonPath": "/paths/~1example/get/responses/200",
        "kind": "response",
      },
      "removed": Object {
        "before": Object {
          "description": "Example response",
          "statusCode": 200,
        },
      },
    },
  ],
  "next": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "operationId": "getExample",
          "responses": Object {},
          "summary": "Retrieve example",
          "tags": Array [
            "Example",
          ],
        },
      },
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "changeType": "removed",
        "location": Object {
          "conceptualLocation": Object {
            "inResponse": Object {
              "statusCode": "200",
            },
            "method": "get",
            "path": "/example",
          },
          "conceptualPath": Array [
            "operations",
            "/example",
            "get",
            "responses",
            "200",
          ],
          "jsonPath": "/paths/~1example/get/responses/200",
          "kind": "response",
        },
        "removed": Object {
          "before": Object {
            "description": "Example response",
            "statusCode": 200,
          },
        },
      },
      "condition": "not be removed",
      "docsLink": undefined,
      "error": "Cannot read property 'statusCode' of undefined",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "removed response: 200",
    },
  ],
}
`;

exports[`operationId fails if changed 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "operationId": "example",
          "responses": Object {},
        },
      },
    },
  },
  "changelog": Array [
    Object {
      "changeType": "changed",
      "changed": Object {
        "after": Object {
          "method": "get",
          "operationId": "example2",
          "pathPattern": "/example",
        },
        "before": Object {
          "method": "get",
          "operationId": "example",
          "pathPattern": "/example",
        },
      },
      "location": Object {
        "conceptualLocation": Object {
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
        ],
        "jsonPath": "/paths/~1example/get",
        "kind": "operation",
      },
    },
  ],
  "next": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "operationId": "example2",
          "responses": Object {},
        },
      },
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "changeType": "changed",
        "changed": Object {
          "after": Object {
            "method": "get",
            "operationId": "example2",
            "pathPattern": "/example",
          },
          "before": Object {
            "method": "get",
            "operationId": "example",
            "pathPattern": "/example",
          },
        },
        "location": Object {
          "conceptualLocation": Object {
            "method": "get",
            "path": "/example",
          },
          "conceptualPath": Array [
            "operations",
            "/example",
            "get",
          ],
          "jsonPath": "/paths/~1example/get",
          "kind": "operation",
        },
      },
      "condition": "have consistent operation IDs",
      "docsLink": undefined,
      "error": "expected 'example' to equal 'example2'",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "updated operation: GET /example",
    },
  ],
}
`;

exports[`operationId fails if removed 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "operationId": "example",
          "responses": Object {},
        },
      },
    },
  },
  "changelog": Array [
    Object {
      "changeType": "changed",
      "changed": Object {
        "after": Object {
          "method": "get",
          "pathPattern": "/example",
        },
        "before": Object {
          "method": "get",
          "operationId": "example",
          "pathPattern": "/example",
        },
      },
      "location": Object {
        "conceptualLocation": Object {
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
        ],
        "jsonPath": "/paths/~1example/get",
        "kind": "operation",
      },
    },
  ],
  "next": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {},
        },
      },
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "changeType": "changed",
        "changed": Object {
          "after": Object {
            "method": "get",
            "pathPattern": "/example",
          },
          "before": Object {
            "method": "get",
            "operationId": "example",
            "pathPattern": "/example",
          },
        },
        "location": Object {
          "conceptualLocation": Object {
            "method": "get",
            "path": "/example",
          },
          "conceptualPath": Array [
            "operations",
            "/example",
            "get",
          ],
          "jsonPath": "/paths/~1example/get",
          "kind": "operation",
        },
      },
      "condition": "have consistent operation IDs",
      "docsLink": undefined,
      "error": "expected 'example' to equal undefined",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "updated operation: GET /example",
    },
  ],
}
`;
